pkg/server/autoconfig/task_markers.go

```diff
@@ -0,0 +1,189 @@
+// licenses/APL.txt.
+
+package autoconfig
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/util/encoding"
+	"github.com/cockroachdb/errors"
+)
+
+// InfoKeyCompletionPrefix is the prefix of the key inserted in job_info
+// when a task has completed.
+var InfoKeyCompletionPrefix = roachpb.Key("completed-")
+
+// InfoKeyStartPrefix is the prefix of the key inserted in job_info
+// when a task has started.
+var InfoKeyStartPrefix = roachpb.Key("started-")
+
+// DecodeTaskID returns the task ID for the given infoKey.
+func DecodeTaskID(prefix roachpb.Key, infoKey []byte) (uint64, error) {
+	if !bytes.HasPrefix(infoKey, []byte(prefix)) {
+		return 0, errors.AssertionFailedf("programming error: prefix %q missing: %q", string(prefix), string(infoKey))
+	}
+	infoKey = infoKey[len(prefix):]
+	_, v, err := encoding.DecodeUvarintAscending(infoKey)
+	return v, err
+}
+
+// MakeTaskInfoKey creates a job_info infoKey for the given task.
+func MakeTaskInfoKey(prefix roachpb.Key, taskID uint64) roachpb.Key {
+	infoKey := make([]byte, 0, len(prefix)+10)
+	infoKey = append(infoKey, []byte(prefix)...)
+	infoKey = encoding.EncodeUvarintAscending(infoKey, taskID)
+	return infoKey
+}
+
+// writeStartMarker writes a start marker for the given task ID and
+// also writes its job ID into the value part.
+func writeStartMarker(
+	ctx context.Context, txn isql.Txn, execCfg *sql.ExecutorConfig, jobID jobspb.JobID, taskID uint64,
+) error {
+	startMarkerInfoKey := MakeTaskInfoKey(InfoKeyStartPrefix, taskID)
+	// We cannot use the InfoStorage directly here because this function
+	// is called from two different jobs (the runner and the task) but
+	// must always write with the job ID of the runner to job_info.
+	_, err := txn.ExecEx(ctx, "write-start-marker", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride,
+		`INSERT INTO system.job_info (job_id, info_key, value) VALUES ($1, $2, $3)`,
+		jobs.AutoConfigRunnerJobID,
+		startMarkerInfoKey,
+		[]byte(strconv.FormatUint(uint64(jobID), 10)))
+	return err
+}
+
+// getCurrentlyStartedTaskID retrieves the ID of the last task which
+// has a start marker in job_info.
+func getCurrentlyStartedTaskID(
+	ctx context.Context, txn isql.Txn, execCfg *sql.ExecutorConfig,
+) (prevTaskID uint64, prevJobID jobspb.JobID, err error) {
+	row, err := txn.QueryRowEx(ctx,
+		"get-last-task-start-marker", txn.KV(), sessiondata.NodeUserSessionDataOverride,
+		// Note: we need the condition "AND info_key < prefixend" so as
+		// to avoid observing other non-marker job info entries.
+		`SELECT info_key, value FROM system.job_info
```

Comment:
Can we use the `Iterate` method on InfoStorage to go over all the rows corresponding to `jobs.AutoConfigRunnerJobID` and with an info_key prefix == InfoKeyStartPrefix?

Reply:
huh no. There could be thousands of them. We are using the index here to get directly to the latest one.

Reply:
Makes sense, thanks